kong/core/router.lua

```diff
@@ -619,30 +621,18 @@ function _M.new(apis)
     end
 
 
-    upstream_host = api_t.upstream_host
-
-
     if api_t.strip_uri_regex then
       local stripped_uri = re_sub(uri, api_t.strip_uri_regex, "/$1", "jo")
       ngx.req.set_uri(stripped_uri)
     end
 
 
-    if api_t.preserve_host then
-      if not headers then
-        headers = ngx.req.get_headers()
-      end
-
-      upstream_host = headers["host"]
-    end
-
-
     if ngx.var.http_kong_debug then
       ngx.header["Kong-Api-Name"] = api_t.api.name
     end
 
 
-    return api_t.api, api_t.upstream_scheme, upstream_host, api_t.upstream_port
+    return api_t.api, api_t.upstream_scheme, api_t.upstream_host, api_t.upstream_port
```

Comment:
For now I kept the list and added the `host_header`, seemed cleanest to me. All header fetching, including checks for `preserve_host` are back in the router.